### ISSUE TITLE
we should always set MODE_RELEASE_BUFFERS on OpenSSL contexts by default

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8247]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#8247|
|Type|enhancement|
|Created|2016-03-23 22:26:46Z|
|Branch|https://github.com/twisted/twisted/tree/8247-tls-mode-release-buffers|

Ben Bangert of Mozilla discovered that TLS connections can be quite memory-intensive within Twisted.  As part of his investigation into why, he discovered that the majority of the overhead can be eliminated with the `SSL_MODE_RELEASE_BUFFERS` flag, documented [here](https://www.openssl.org/docs/manmaster/ssl/SSL_CTX_set_mode.html) as:

> When we no longer need a read buffer or a write buffer for a given SSL, then release the memory we were using to hold it. Using this flag can save around 34k per idle SSL connection. This flag has no effect on SSL v2 connections, or on DTLS connections.

Given that Python copies every string like a zillion times anyway, and modern malloc implementations are quite fast, I doubt that freeing these buffers would be significant CPU overhead; however, it might save significant amounts of memory in concurrent servers.  We should always set it.

It is exported by PyOpenSSL as `OpenSSL.SSL.MODE_RELEASE_BUFFERS`.  I am not sure about what versions may or may not support it.

<details><summary>Searchable metadata</summary>

```
trac-id__8247 8247
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__None None
branch__8247_tls_mode_release_buffers 8247-tls-mode-release-buffers
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1458772006597862 1458772006597862
changetime__1479797924590355 1479797924590355
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban commented|
|-|-|

In OpenSSL v0.13 it is not available.

It was introduced here https://github.com/pyca/pyopenssl/commit/a63714c6ef6915b662a5fa53094fd8c6e3aa9be8 ... or 0.14

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

PR at https://github.com/twisted/twisted/pull/602

It does not do the OpenSSL checking that Python 3.5+ does, it's been long enough that all supported OSs with those versions should have patched it (and you can bring your own OpenSSL with Twisted, too).

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_4"></a>|@Lukasa set owner to @hawkowl|
|-|-|

So, my position on those checks is complex. I think in general the checks are a good idea: older OpenSSL versions most certainly do still float around, and exposing people to CVEs is bad.

On the other hand, there's a solid school of thought that says that if you're still running a broken OpenSSL then you're in a world of pain of your own making. Given that having such a branch would be annoying to test, I think I'm ok with not providing the checks.

Please resolve the txchecker failures and merge. :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set status to `closed`|
|-|-|

In changeset 7314e23b638ae2f59cce91107c41bd4567789044
```
#!CommitTicketReference repository="" revision="7314e23b638ae2f59cce91107c41bd4567789044"
Merge 8247-tls-mode-release-buffers: Set MODE_RELEASE_BUFFERS in CertificateOptions

Author: hawkowl
Reviewer: lukasa
Fixes: #8247
```
